core/commit.go

```diff
@@ -154,15 +154,23 @@ func makeCommitmentCounter(f uint32) commitmentCounter {
 		if replicasCommitted[replicaID] {
 			return false, fmt.Errorf("Duplicated commitment")
 		}
-		replicasCommitted[replicaID] = true
 
-		if len(replicasCommitted) == int(f+1) {
-			delete(prepareStates, prepareCV)
-			lastDoneCV++
+		for cv := lastDoneCV + 1; cv < prepareCV; cv++ {
+			s := prepareStates[cv]
+			if len(s) != 0 && !s[replicaID] {
+				return false, fmt.Errorf("Skipped commitment")
+			}
+		}
 
-			return true, nil
+		if len(replicasCommitted) < int(f) {
+			replicasCommitted[replicaID] = true
+
+			return false, nil
 		}
 
-		return false, nil
+		lastDoneCV = prepareCV
```

Comment:
This implicitly allows us to increase `lastDoneCV` by more than 1. If that should be considered as error in any possible case, putting some assertion to detect that might be desirable.

Reply:
If not the CHECKPOINT message which also gets UI assigned, the PREPARE/COMMIT messages would always have sequential UIs. So, according to the paper, this is not an error.